Bug 30967 - ls from nfs4 mount path from is slow even for 10k files f…

SriKrishnachowdaryk opened the pull request:
…rom user.

Problem description:
1. nfs ganesha does readdir and fetches all directory entries into dirent cache
2. It invalidates dirent cache every 3 seconds.
3. In one readdir operation client can consume a maximum of 50 entries at most.
4. If by the time client consumed these entries, 3 seconds are passes, the readdir in step (1) is re-attempted.
5. This means for 50 entries, nfs4, through mapr client endsup making 100s of readdir RPC, 1000s of lookup RPC, 1000s of getattr RPC.

All these were creating the slow down.

So, the fix targets reducing the number of RPC by following steps

1. Allow nfs4 server to issue a readdirplus request
2. Get the fsal_obj_handle through the readdir callback and avoid lookups
3. Specify the cookie and num_entries that nfs4 server can server to minimize readdir overhead
4. Skip direnct cache and exit readdirplus as soon as num_entries are populated.
5. Make sure for large directories other operations like create, remove, rename do not use dirent cache.
6. Also Made optimizations in the fsal layer to avoid getattrs calls.

satgs commented on src/cache_inode/cache_inode_readdir.c:
```diff
@@ -494,13 +553,25 @@ cache_inode_readdir_populate(cache_entry_t *directory)
 	state.directory = directory;
 	state.status = &status;
 	state.offset_cookie = 0;
+	state.nfs4_cb_parms = cb_parms;
```
Add a note : These are needed for getattr

satgs commented on src/cache_inode/cache_inode_readdir.c:
```diff
@@ -494,13 +553,25 @@ cache_inode_readdir_populate(cache_entry_t *directory)
+	state.prev_obj_hdl_init = 0;
+
+	/* read entire directory if directory can be cacheable */
+	if (directory->flags & CACHE_INODE_DIRENT_AVL_TREE) {
```
fix indentation.

satgs replied:
print a log message here and in the else path.

satgs commented on src/cache_inode/cache_inode_readdir.c:
```diff
@@ -660,6 +738,18 @@ cache_inode_readdir(cache_entry_t *directory,
 
+	/* directory's entries are not cached */
+	if (!(directory->flags & CACHE_INODE_DIRENT_AVL_TREE)) {
+		LogDebug(COMPONENT_NFS_READDIR,
```
Print the ino.

SriKrishnachowdaryk replied:
I could not find ino. Neither could I find the name Satya.

satgs commented on src/include/cache_inode.h:
```diff
@@ -1042,6 +1047,17 @@ cache_inode_refresh_attrs(cache_entry_t *entry)
+
+	/* check filesize and cache small files */
+	if ((entry->type == DIRECTORY) &&
+	    (entry->obj_handle->attrs->filesize <= nfs_param.nfsv4_param.dirent_cache_threshold)) {
```
filesize >=  nfs_param.nfsv4_param.dirent_cache_threshold

satgs commented on src/cache_inode/cache_inode_misc.c:
```diff
@@ -822,6 +834,14 @@ cache_inode_release_dirents(cache_entry_t *entry,
+					     "hdl:%p directory entries:%lu, dirent_cache_threshold:%lu",
+					     entry->obj_handle, entry->obj_handle->attrs->filesize,
+					     nfs_param.nfsv4_param.dirent_cache_threshold);
+				entry->flags &= ~CACHE_INODE_DIRENT_AVL_TREE;
```
Setting code should go here.

satgs commented on src/cache_inode/cache_inode_misc.c:
```diff
@@ -822,6 +834,14 @@ cache_inode_release_dirents(cache_entry_t *entry,
 			entry->object.dir.nbactive = 0;
 			atomic_clear_uint32_t_bits(&entry->flags,
 						   CACHE_INODE_DIR_POPULATED);
+			if (entry->obj_handle->attrs->filesize >
```
1. You should blindly delete all entries. Do not look at the cache flag.
2. You should delay setting cache flag by 3 seconds. 
3. Check if you hold the right lock to modify entry.

SriKrishnachowdaryk replied:
atta_lock and content_lock are protecting the file size from being modified while we do readdir.

satgs reviewed (commented):
Incorporate changes and merge the code.

Commit:
Bug 30967 - ls from nfs4 mount path from is slow even for 10k files from user

skipBuild
unitTested

Problem description:
1. nfs ganesha does readdir and fetches all directory entries into dirent cache
2. It invalidates dirent cache every 3 seconds.
3. In one readdir operation client can consume a maximum of 50 entries at most.
4. If by the time client consumed these entries, 3 seconds are passes, the readdir in step (1) is re-attempted.
5. This means for 50 entries, nfs4, through mapr client endsup making 100s of readdir RPC, 1000s of lookup RPC, 1000s of getattr RPC.

All these were creating the slow down.

So, the fix targets reducing the number of RPC by following steps

1. Allow nfs4 server to issue a readdirplus request
2. Get the fsal_obj_handle through the readdir callback and avoid lookups
3. Specify the cookie and num_entries that nfs4 server can server to minimize readdir overhead
4. Skip dirent cache and exit readdirplus as soon as num_entries are populated.
5. Make sure for large directories other operations like create, remove, rename do not use dirent cache.
6. Populate attrs from readdirplus results and update with entries in cache.

Change-Id: Iebf1c9746c918f306a11d94cc1bd540a6e5b9907
Signed-off-by: Krishna Kankanala <kkankanala@maprtech.com>
